### PR TITLE
kubeadm: make "alpha kubeconfig user" accept --config

### PR DESCRIPTION
**Pull request by knight42**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

xref https://github.com/kubernetes/kubeadm/issues/2292

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: make the command "kubeadm alpha kubeconfig user" accept a "--config" flag and remove the following flags:
- apiserver-advertise-address / apiserver-bind-port: use either localAPIEndpoint from InitConfiguration or controlPlaneEndpoint from ClusterConfiguration.
- cluster-name: use clusterName from ClusterConfiguration
- cert-dir: use certificatesDir from ClusterConfiguration
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/alpha/kubeconfig.go [hunk 1]
```diff
@@ -81,29 +77,23 @@ func newCmdUserKubeConfig(out io.Writer) *cobra.Command {
...
 	}
 
-	// Add ClusterConfiguration backed flags to the command
-	cmd.Flags().StringVar(&clusterCfg.CertificatesDir, options.CertificatesDir, clusterCfg.CertificatesDir, "The path where certificates are stored")
-	cmd.Flags().StringVar(&clusterCfg.ClusterName, "cluster-name", clusterCfg.ClusterName, "Cluster name to be used in kubeconfig")
-
-	// Add InitConfiguration backed flags to the command
-	cmd.Flags().StringVar(&initCfg.LocalAPIEndpoint.AdvertiseAddress, options.APIServerAdvertiseAddress, initCfg.LocalAPIEndpoint.AdvertiseAddress, "The IP address the API server is accessible on")
-	cmd.Flags().Int32Var(&initCfg.LocalAPIEndpoint.BindPort, options.APIServerBindPort, initCfg.LocalAPIEndpoint.BindPort, "The port the API server is accessible on")
+	options.AddConfigFlag(cmd.Flags(), &cfgPath)
```
Comment (knight42) [inline cmd/kubeadm/app/cmd/alpha/kubeconfig.go:96]: @neolit123 Currently this command only accept local kubeadm config, should we also consider the config saved in cluster?

Comment (neolit123) [reply cmd/kubeadm/app/cmd/alpha/kubeconfig.go:96]: thanks for the PR, i can have a deeper look next week.

that is a good question. we had previous discussions whether certain kubeadm commands should use the kubeadm-config CM. my take is that if a command has potential to *not operate* on the running cluster it should not fetch the CM.

for example, if the command is currently being run on a Node part of a cluster and if a kubeconfig to access the cluster is in the user home dir, the user might still want to create a kubeadm for a completely different cluster, but the command will use the configuration from the cluster, which might not be desired.

i had the same argument for the command "kubeadm config images...". "kubeadm alpha certs renew" on the other hand has CM fetch, but the idea of the command is to renew certs for this Node / cluster.

so my preference here would be the following:
- the user should feed --config and the command should fail without it.
- if the user wants to feed the --config for the current cluster they can fetch the ClusterConfiguration from the CM.

Comment (knight42) [reply cmd/kubeadm/app/cmd/alpha/kubeconfig.go:96]: I got it, thanks for you detailed clarification!

**Comment by neolit123** [comment]
doing a minor retitle:

/retitle kubeadm: make "alpha kubeconfig user" accept --config

**Comment by neolit123** [comment]
please change the release note to:

```
kubeadm: make the command "kubeadm alpha kubeconfig user" accept a "--config" flag and remove the following flags:
- apiserver-advertise-address / apiserver-bind-port: use either localAPIEndpoint from InitConfiguration or controlPlaneEndpoint from ClusterConfiguration.
- cluster-name: use clusterName from ClusterConfiguration
- cert-dir: use certificatesDir from ClusterConfiguration
```

File: cmd/kubeadm/app/cmd/alpha/kubeconfig.go [hunk 2]
```diff
@@ -79,36 +75,35 @@ func newCmdUserKubeConfig(out io.Writer) *cobra.Command {
...
 
 	// Add command specific flags
 	cmd.Flags().StringVar(&token, options.TokenStr, token, "The token that should be used as the authentication mechanism for this kubeconfig, instead of client certificates")
 	cmd.Flags().StringVar(&clientName, "client-name", clientName, "The name of user. It will be used as the CN if client certificates are created")
 	cmd.Flags().StringSliceVar(&organizations, "org", organizations, "The orgnizations of the client certificate. It will be used as the O if client certificates are created")
 
+	_ = cmd.MarkFlagRequired(options.CfgPath)
+	_ = cmd.MarkFlagRequired("client-name")
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/alpha/kubeconfig.go:107]: these can be without the leading `_ =`, unless linters complain?

File: cmd/kubeadm/app/cmd/alpha/kubeconfig.go [hunk 3]
```diff
@@ -79,36 +75,35 @@ func newCmdUserKubeConfig(out io.Writer) *cobra.Command {
 			if clientName == "" {
 				return errors.New("missing required argument --client-name")
 			}
+			if cfgPath == "" {
+				return errors.New("missing required argument --config")
+			}
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/alpha/kubeconfig.go:80]: we can remove the above `if` checks if cmd.MarkFlagRequired is used.

File: cmd/kubeadm/app/cmd/alpha/kubeconfig_test.go [hunk 4]
```diff
@@ -19,16 +19,66 @@ package alpha
...
+			APIVersion: "kubeadm.k8s.io/v1beta2",
+			Kind:       "InitConfiguration",
+		},
+		LocalAPIEndpoint: kubeadmapiv1beta2.APIEndpoint{
+			AdvertiseAddress: "1.2.3.4",
+			BindPort:         1234,
+		},
+		NodeRegistration: kubeadmapiv1beta2.NodeRegistrationOptions{
+			CRISocket: "/path/to/dockershim.sock",
+		},
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/alpha/kubeconfig_test.go:53]: i think the NodeRegistration above can be removed

File: cmd/kubeadm/app/cmd/alpha/kubeconfig_test.go [hunk 5]
```diff
@@ -19,16 +19,66 @@ package alpha
...
+	"sigs.k8s.io/yaml"
+
+	kubeadmapiv1beta2 "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2"
 	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
 	"k8s.io/kubernetes/cmd/kubeadm/app/util/pkiutil"
 	testutil "k8s.io/kubernetes/cmd/kubeadm/test"
 	kubeconfigtestutil "k8s.io/kubernetes/cmd/kubeadm/test/kubeconfig"
 )
 
+func generateKubeadmConfig(dir, id, certDir, clusterName string) (string, error) {
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/alpha/kubeconfig_test.go:39]: possibly best to indicate this is used for tests package scope wise - e.g. name the function generateTestKubeadmConfig

File: cmd/kubeadm/app/cmd/alpha/kubeconfig_test.go [hunk 6]
```diff
@@ -19,16 +19,66 @@ package alpha
...
+	}
+	buf.Write(data)
+	buf.WriteString("---\n")
+	data, err = yaml.Marshal(&clusterCfg)
+	if err != nil {
+		return "", err
+	}
+	buf.Write(data)
+
+	err = ioutil.WriteFile(cfgPath, buf.Bytes(), 0o644)
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/alpha/kubeconfig_test.go:78]: should be `0644` for consistency with the rest of kubeadm code.

File: cmd/kubeadm/app/cmd/alpha/kubeconfig_test.go [hunk 7]
```diff
@@ -19,16 +19,66 @@ package alpha
...
+
+	kubeadmapiv1beta2 "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2"
 	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
 	"k8s.io/kubernetes/cmd/kubeadm/app/util/pkiutil"
 	testutil "k8s.io/kubernetes/cmd/kubeadm/test"
 	kubeconfigtestutil "k8s.io/kubernetes/cmd/kubeadm/test/kubeconfig"
 )
 
+func generateKubeadmConfig(dir, id, certDir, clusterName string) (string, error) {
+	id = strings.ReplaceAll(id, " ", "_") + ".conf"
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/alpha/kubeconfig_test.go:40]: i think we can leave "id" here as is, without replacing " " and adding the extension.

File: cmd/kubeadm/app/cmd/alpha/kubeconfig.go [hunk 8]
```diff
@@ -62,11 +62,7 @@ func newCmdUserKubeConfig(out io.Writer) *cobra.Command {
 	initCfg := &kubeadmapiv1beta2.InitConfiguration{}
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/alpha/kubeconfig.go:62]: we should update the above example too:
```
userKubeconfigExample = cmdutil.Examples(`
	# Output a kubeconfig file for an additional user named foo using a kubeadm config file bar
	kubeadm alpha kubeconfig user --client-name=foo --config=bar
	`)
```

**Review by neolit123 — COMMENTED** [review]
@knight42 thanks for the PR, i've added some minor comments.
i think we can keep the commits squashed to 1 here.

LMK if you have any questions.

File: cmd/kubeadm/app/cmd/alpha/kubeconfig.go (hunk 2 above)
Comment (knight42) [reply cmd/kubeadm/app/cmd/alpha/kubeconfig.go:107]: yeah just want to make goland happy.

File: cmd/kubeadm/app/cmd/alpha/kubeconfig.go [hunk 9]
```diff
@@ -40,7 +39,7 @@ var (
 
 	userKubeconfigExample = cmdutil.Examples(`
 	# Output a kubeconfig file for an additional user named foo
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/alpha/kubeconfig.go:41]: probably best to indicate that --config is the kubeadm configuration, as mentioned here:
https://github.com/kubernetes/kubernetes/pull/94879/files#r492257006

**Review by neolit123 — COMMENTED** [review]
added one final comment. i think this is good to go after that.
/approve

**Comment by neolit123** [comment]
thank you @knight42 
if someone objects to the change made here we can amend before 1.20 is released.
/lgtm
/milestone v1.20

**Pull request merged**